specs/AddHostObjectToScriptWithOrigins.md

```diff
@@ -0,0 +1,282 @@
+# Background
+Currently WebView2 supports adding host objects to script from client application
+and then later those objects can be accessed from the JavaScript of the top level frame
+using `window.chrome.webview.hostObjects.{name}`.
+Documentation can be found [here](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2?view=webview2-1.0.705.50#addhostobjecttoscript).
+
+We were asked to support this feature in iframes including
+child frames that don't match the origin of the parent. In this document we describe API
+additions to support host objects in iframes. We'd appreciate your feedback.
+
+
+# Description
+The FrameCreated event is raised when an iframe is created and before starting any
+navigation and content loading. The FrameDeleted event is raised when an iframe is deleted
+or the document containing that iframe is destroyed.
+By providing event handler for frame created event,
+user can obtain the iframe and add host objects to it specifying the list of origins
+using `AddHostObjectToScriptWithOrigins` method.
+When script in the iframe tries to access a host object the origin of the iframe will be
+checked against the provided list. The Frame object has a Name property with the value of the
+name attribute from the iframe html tag declaring it. The NameChanged event is raised when
+corresponding iframe changes its window.name property.
+
+The host object will be removed during iframe deletion automatically.
+
+
+# Examples
+## Frame created event, frame name changed event, adding and removing host object
+
+## .NET, WinRT
+```c#
+void SubscribeToFrameCreated()
+{
+    webView.CoreWebView2.FrameCreated += delegate (object sender, CoreWebView2FrameCreatedEventArgs args)
+    {
+        try
+        {
+            if (args.Frame.Name.Equals("iframe_name"))
+            {
+                String[] origins = new String[] { "https://appassets.example" };
+                args.Frame.AddHostObjectToScriptWithOrigins("bridge", new BridgeAddRemoteObject(), origins);
+            }
+            args.Frame.NameChanged += delegate (object nameChangedSender, object nameChangedArgs)
+            {
+                CoreWebView2Frame frame = (CoreWebView2Frame)nameChangedSender;
+                if (!frame.Name.Equals("iframe_name"))
+                {
+                    frame.RemoveHostObjectFromScript("bridge");
+                }
+            };
+        }
+        catch (NotSupportedException exception)
+        {
+            MessageBox.Show("Frame.AddHostObjectToScriptWithOrigins failed: " + exception.Message);
```

Comment:
We want to avoid running MessageBox in our event handlers or async method completion handlers in our samples since MessageBox runs its own message loop and can lead to confusing reentrancy. But this is for NotSupportedException which shouldn't really come up for the experimental or release version of this API correct? We can remove this try/catch for the sample code since the end dev shouldn't be able to see a NotSupportedException?

Reply:
In sample app this part is wrapped into `#if USE_WEBVIEW2_STAGING`, and it crashes without catching exception in Debug configuration. Will rebuild for Release and update with results.

Reply:
Ok, I reread the initial message - try-catch is indeed for NotSupportedException  and because flag USE_WEBVIEW2_STAGING is not set in public sample app(event it is set in Release private one) I have removed it here.